lib/controllers/category_controller.js

```diff
@@ -0,0 +1,54 @@
+let CategoryModel = require('../models/city_model')
+
+function Category () { }
+
+// CREATE CATEGORY
+Category.prototype.createCategoryForCity = function (cityId, catObj, callback) {
+  if (!cityId || !catObj) return callback(new Error('Something went wrong'))
+  if (catObj.category_name) catObj.type_of = catObj.category_name.toLowerCase()
+  return CategoryModel.findOne({ _id: cityId }, function (err, result) {
```

Comment:
Whats the right way to use callbacks in this case? When I need to do two async operations in one.

Reply:
In this case I'd take a look at `async.waterfall` https://caolan.github.io/async/docs.html#waterfall

Or! If you wanna be cool use the `async/await` JS feature (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function). For that you'd have to _not_ use callbacks, but use Mongoose's promises (https://mongoosejs.com/docs/promises.html). Which, btw, you should be doing anyway. If you ever can not use callbacks and promises are an option - _definitely_ use promises. Promises can do things that callbacks cannot, like deferring the handling.

With promises:
```
Category.prototype.createCategoryForCity = {
  return CategoryModel.findOne({ _id: cityId }).then(function(err, result) {
   ...
        return CategoryModel.findOneAndUpdate(
          { _id: cityId },
          { $push: { categories: catObj } },
          { new: true },
        )
  })
})

// in router
router.post('/:city_id/categories', (req, res) => {
    Category.createCategoryForCity(req.params.city_id, req.body).then((err, result) => {
    ...
    })
    ...
})
```

Have not used await/async much. So would have to think a little about error handling and stuff... But it's something like this
```
Category.prototype.createCategoryForCity = async function() {
    const category = await CategoryModel.findOne({ _id: cityId })
    if (!category) {
      throw Error()
    }

    catObj.order = result.categories.length + 1

    return CategoryModel.findOneAndUpdate(
      { _id: cityId },
      { $push: { categories: catObj } },
      { new: true },
    )
})

// in router
router.post('/:city_id/categories', (req, res) => {
    Category.createCategoryForCity(req.params.city_id, req.body).then((err, result) => {
    ...
    })
    ...
})
```

Reply:
But doing one thing at a time it not a bad idea. So just learn to use promises all the time first. You can figure out `async/await` later.